### PR TITLE
IGNITE-12853: Introduced features for the thin client protocol

### PR DESCRIPTION
**Pull request by isapego**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientUtils.java [hunk 1]
```diff
@@ -352,7 +351,7 @@ void cacheConfiguration(ClientCacheConfiguration cfg, BinaryOutputStream out, Pr
                 )
             );
 
-            if (ver.compareTo(V1_6_0) >= 0) {
+            if (protocolContext.isExpirationPolicySupported()) {
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientUtils.java:354]: `ExpirationPolicy` -> `ExpiryPolicy`, I think that's the term we use in Ignite

File: modules/platforms/cpp/thin-client/src/impl/data_channel.h [hunk 2]
```diff
@@ -59,10 +59,22 @@ namespace ignite
...
                 /** Version 1.4.0. Added: Partition awareness support, IEP-23. */
                 static const ProtocolVersion VERSION_1_4_0;
 
+                /** Version 1.5.0. Transaction support. */
+                static const ProtocolVersion VERSION_1_5_0;
+
+                /** Version 1.6.0. Expiration Policy Configuration. */
+                static const ProtocolVersion VERSION_1_6_0;
+
+                /** Version 1.6.0. Features introduced. */
```
Comment (ptupitsyn) [inline modules/platforms/cpp/thin-client/src/impl/data_channel.h:75]: 1.7.0

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.EnumSet;
+
+/**
+ * The base feature class.
+ */
+public interface ThinProtocolFeature {
+    /**
+     * @return Feature ID.
+     */
+    int featureId();
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java:31]: "public" modificator must be present on all interface methods (codestyle). But I think the whole interface should be removed (see my comment on ticket for details)

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * @return Feature name.
+     */
+    String name();
+
+    /**
+     * @param features Feature set.
+     * @return Byte array representing all supported features by current node.
+     */
+    static <E extends Enum<E> & ThinProtocolFeature> byte[] featuresAsBytes(E[] features) {
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java:42]: The method is never used

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java [hunk 5]
```diff
@@ -520,9 +520,10 @@ private void writeCacheInfo(PayloadOutputChannel payloadCh) {
         TcpClientTransaction tx = transactions.tx();
 
         if (expiryPlc != null) {
-            if (payloadCh.clientChannel().serverVersion().compareTo(V1_6_0) < 0) {
-                throw new ClientProtocolError(String.format("Expire policies have not supported by the server " +
-                    "version %s, required version %s", payloadCh.clientChannel().serverVersion(), V1_6_0));
+            ProtocolContext protocolContext = payloadCh.clientChannel().protocolContext();
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java:523]: Newline missed (codestyle)

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientProtocolContext.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+package org.apache.ignite.internal.processors.platform.client;
+
+import java.util.EnumSet;
+import org.apache.ignite.internal.ThinProtocolFeature;
+import org.apache.ignite.internal.processors.odbc.ClientListenerProtocolVersion;
+
+import static org.apache.ignite.internal.processors.platform.client.ClientConnectionContext.*;
+
+/**
+ * Protocol context for JDBC protocol. Holds protocol version and supported features.
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientProtocolContext.java:27]: `JDBC` -> `thin client`?

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Configuration/QueryEntity.cs [hunk 7]
```diff
@@ -229,12 +229,31 @@ string IQueryEntityInternal.GetAlias(string fieldName)
...
+        {
+            Read(reader, srvVer);
+        }
+        
+        /// <summary>
+        /// Reads an instance of the <see cref="QueryEntity"/> class.
+        /// </summary>
+        /// <param name="reader">The reader.</param>
+        /// <param name="srvVer">Server version.</param>
+        internal void Read(IBinaryRawReader reader, ClientProtocolVersion srvVer)
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Cache/Configuration/QueryEntity.cs:256]: Ideally we should get rid of `ClientProtocolVersion` here and separate Thin and Thick logic completely. Thin client config logic should be in `ClientCacheConfigurationSerializer`.

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientUtils.java [hunk 8]
```diff
@@ -366,8 +366,8 @@ void cacheConfiguration(ClientCacheConfiguration cfg, BinaryOutputStream out, Pr
                 });
             }
             else if (cfg.getExpiryPolicy() != null) {
-                throw new ClientProtocolError(String.format("Expire policies have not supported by the server " +
-                    "version %s, required version %s", ver, V1_6_0));
+                throw new ClientProtocolError(String.format("Expire policies are not supported by the server " +
+                    "version %s, required version %s", protocolCtx.version(), V1_6_0));
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientUtils.java:370]: `V1_6_0` -> `ProtocolVersionFeature.EXPIRY_POLICY.verIntroduced()`

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java [hunk 9]
```diff
@@ -520,9 +520,11 @@ private void writeCacheInfo(PayloadOutputChannel payloadCh) {
...
 
         if (expiryPlc != null) {
-            if (payloadCh.clientChannel().serverVersion().compareTo(V1_6_0) < 0) {
-                throw new ClientProtocolError(String.format("Expire policies have not supported by the server " +
-                    "version %s, required version %s", payloadCh.clientChannel().serverVersion(), V1_6_0));
+            ProtocolContext protocolCtx = payloadCh.clientChannel().protocolCtx();
+
+            if (!protocolCtx.isFeatureSupported(ProtocolVersionFeature.EXPIRY_POLICY)) {
+                throw new ClientProtocolError(String.format("Expire policies are not supported by the server " +
+                    "version %s, required version %s", protocolCtx.version(), V1_6_0));
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java:527]: `V1_6_0` -> `ProtocolVersionFeature.EXPIRY_POLICY.verIntroduced()`

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientChannel.java [hunk 10]
```diff
@@ -446,39 +463,46 @@ private void handshakeRes(String user, String pwd, Map<String, String> userAttrs
...
+                    errCode = reader.readInt();
 
                 if (errCode == ClientStatus.AUTH_FAILED)
                     throw new ClientAuthenticationException(err);
-                else if (ver.equals(srvVer))
+                else if (proposedVer.equals(srvVer))
                     throw new ClientProtocolError(err);
                 else if (!supportedVers.contains(srvVer) ||
-                    (srvVer.compareTo(V1_1_0) < 0 && user != null && !user.isEmpty()))
+                    (srvVer.compareTo(V1_1_0) < 0 && !F.isEmpty(user)))
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientChannel.java:495]: `srvVer.compareTo(V1_1_0) < 0` -> `!ProtocolContext.isFeatureSupported(srvVer, ProtocolVersionFeature.AUTHORIZATION)`

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientTransactions.java [hunk 11]
```diff
@@ -88,13 +88,15 @@ private ClientTransaction txStart0(TransactionConcurrency concurrency, Transacti
...
         tx0 = ch.service(ClientOperation.TX_START,
             req -> {
-                if (req.clientChannel().serverVersion().compareTo(V1_5_0) < 0) {
-                    throw new ClientProtocolError(String.format("Transactions have not supported by the server's " +
-                        "protocol version %s, required version %s", req.clientChannel().serverVersion(), V1_5_0));
+                ProtocolContext protocolCtx = req.clientChannel().protocolCtx();
+
+                if (!protocolCtx.isFeatureSupported(ProtocolVersionFeature.TRANSACTIONS)) {
+                    throw new ClientProtocolError(String.format("Transactions are not supported by the server's " +
+                        "protocol version %s, required version %s", protocolCtx.version(), V1_5_0));
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientTransactions.java:99]: `V1_5_0` -> `ProtocolVersionFeature.TRANSACTIONS.verIntroduced()`

**Pull request closed without merging**